### DISCUSSION TITLE
Concurrency issues?

### DISCUSSION THREAD

**Discussion by wonderplayer** [pnp/pnpcore#611] [General] [answered]
Hello,
I'm working with PnPCore in Azure Function App, and when load testing I notice that requests are failing. It happens when I'm executing custom API batch requests.
I'm not entirely sure where the issue is, but when Azure queue messages are processed in smaller batches and there are less parallel instances that execute requests, then everything seems fine.
I will try to dig a bit into this and try to catch at least error message.

**Comment by jansenbe** [comment, marked as answer]
@wonderplayer : the `PnPContext` class is not thread safe, so ensure you create a new one per thread. If you can capture an error and share the stacktrace than that will help with understanding what goes wrong.

**Reply by wonderplayer** [reply]
Maybe you have some suggestion how can I catch batch error response?
I'm trying to catch `response.IsSuccessStatusCode == false`, but it doesn't get hit.
![image](https://user-images.githubusercontent.com/15668343/140486251-54c74b43-b9f9-404d-ac4f-06dcdb651cd9.png)

**Reply by wonderplayer** [reply]
I've added simple information to requests with thread ID and see that actually different threads in the same process are used when creating PnPContext.

**Reply by wonderplayer** [reply]
I see that last request in batch needs execution, and others not.
Tried to do without batching - a bit better. Still some requests fail, but there are less of them.

**Reply by wonderplayer** [reply]
Exception message that I can get with try/catch patter as described in documentation is following:
`Message: An unexpected error occurred while manipulating the navigational structure of this Web.`

**Reply by wonderplayer** [reply]
Stack trace:
```
   at PnP.Core.Services.BatchClient.ProcessSharePointRestBatchResponseContent(Batch batch, String batchResponse)
   at PnP.Core.Services.BatchClient.<ProcessSharePointRestBatchResponse>d__48.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestBatchAsync>d__46.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at PnP.Core.Services.BatchClient.<ExecuteBatch>d__36.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PnP.Core.Services.PnPContext.<ExecuteAsync>d__67.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at contoso.SiteConfigurationNET.Handlers.SharePointOnlineHandler.<CreateNavigationNodeLevel>d__9.MoveNext() in C:\\Users\\source\\repos\\contoso\\SiteConfigurationNET\\Handlers\\SharePointOnlineHandler.cs:line 105
```

**Reply by wonderplayer** [reply]
Strange thing is that when batch with 1 request failed, there were 2 exactly the same error responses.
![image](https://user-images.githubusercontent.com/15668343/140499689-563bd008-570e-4335-9864-aa4683bda766.png)

**Reply by wonderplayer** [reply]
Never mind, for any batch error there are two exactly same error messages. Even if batch contained 8 requests and one of them failed, error response will still have 2 exactly same error results.

**Reply by jansenbe** [reply]
@wonderplayer : the error is a SPO REST error, so something goes wrong with the requests you make. By default if one of the requests in a batch fails an exception is thrown, but you do can change that behavior by passing the `throwOnError` flag as `false` when calling `ExecuteAsync`. See https://pnp.github.io/pnpcore/using-the-sdk/basics-batching.html#handling-batch-failures for our documentation on this feature.

**Reply by wonderplayer** [reply]
Yes, that's exactly what I did to get error message. I have no idea why this error message appears there, as with some retires everything goes well. And as my batches create child nodes under one parent node, it's very strange that some requests succeed, and some don't.

**Reply by wonderplayer** [reply]
Took a look at [Azure Functions sample](https://github.com/pnp/pnpcore/blob/dev/samples/Demo.AzureFunction/GetSiteDocuments.cs), and added `using` to my code, but still am getting `REST Service exception`. There are only two messages being processed in parallel in different threads, so am not sure, where the issue could be.

**Reply by jansenbe** [reply]
Do both threads use their own PnPContext instance?

**Reply by wonderplayer** [reply]
I have following setup:
Message triggers Azure Function which receives IPnPContextFactory by Dependency Injection. Then I pass this factory to a class, which creates context with CreateAsync() method.
I believe that every message is using it's own PnPContext instance. Is there a way to make sure of that, like context ID or something?

**Reply by jansenbe** [reply]
@wonderplayer : your approach should be fine, you'll have different contexts. There's a context Id property, but it's internal.

Given the error is still an SPO REST one I still think somehow the issues requests are not always valid. Would recommend enabling Debug logging and investigate the requests made (the actual request details will show up in the log when setting log level to Debug). Might it be that there's a timing issue or so, e.g. trying to add a sub nav node while the parent nav node add was not processed?

**Reply by wonderplayer** [reply]
OK, will try to see stuff with Debug.
Batch includes multiple requests and only last request fails, while other requests are in executed state and have success status code. So, parent node is present.

**Reply by jansenbe** [reply]
@wonderplayer : did have a quick peek at our source code and the error you see is not immediately pointing to something obvious. If you can share the correlationid and tenant name of a recent failing request then I can grab the internal ULS logs and see if that can provide more clarity. If you have a "simple" repro that would be great as well, might be that this is a bug

**Reply by wonderplayer** [reply]
Alright, I will create a simple Azure Function App project and replicate this behaviour.
Will add the URL tomorrow 😊

**Reply by jansenbe** [reply]
@wonderplayer : ideally the issue is reproducible outside of an Azure function app (to make repro easier). Maybe start with sending me a correlationid and tenant name (so contoso.sharepoint.com) of a failing request. Feel free to send to bjansen@microsoft.com so your tenant name is not exposed here.

**Reply by jansenbe** [reply]
@wonderplayer : we've had an offline discussion and seems like there's a SharePoint bug being hit when updating (hub) navigation settings. As this is not related to PnP Core SDK threading I'll mark this discussion as answered.
